doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+To represent what Materialize knows with certainty, Materialize maintains two "frontiers" for each TVC.
+A frontier is a lower bound on times, and is used to describe the set of times greater or equal to the lower bound.
+The two frontiers Materialize maintains for each TVC are:
+
+*   The `since` frontier, which communicates that for times `time` greater or equal to `since`, the accumulated collection at `time` is precise.
+
+    Informally, updates at times not greater than the `since` frontier may have been consolidated with other updates.
+    They are not discarded, but their times may have been advanced to the `since` frontier to allow cancelation and consolidation of updates to occur.
+    This is analogous to "version vacuuming" in multiversioned databases.
+    As the system operates, the `since` frontier may advance to allow compaction of data.
+    This happens subject to constraints, and only when correctness permits.
+
+    We think of `since` as the "read frontier": times greater or equal to it may still be correctly read from the TVC.
+
+*   The `upper` frontier, which communicates that for times `time` not greater or equal to `upper`, the updates at `time` are complete.
+
+    The upper frontier provides the information about how a TVC may continue to evolve.
+    All future updates will necessarily have their `time` greater or equal to the `upper` frontier.
+    As the system operates, we expect the `upper` frontiers to advance indicating that the contents of the collection are known for more times.
+    This happens only when we become certain that updates for a time are complete.
+
+    We think of `upper` as the "write frontier": times greater or equal to it may still be written to the TVC.
+
+These two frontiers mean that for any `time` greater or equal to `since`, but not greater or equal to `upper`, we can exactly reconstruct the collection at `time`.
+If `time` is not greater or equal to `since`, we may have permanently lost historical distinctions that allow us to know the contents at `time`.
+If `time` is greater or equal to `upper`, we may still receive changes to the collection that occur at `time`.
+The two frontiers never move backward, and often move forward as the system operates.
+
+At each moment, Materialize's representation of a TVC is as a set of update triples `(data, time, diff)` and the two frontiers `since` and `upper`.
+We call this pair of set of updates and `(since, upper)` frontiers a *partial TVC* or pTVC.
+
+## Evolution
+
+Materialize maintains a (partial) map from `GlobalId`s to pTVCs.
+A `GlobalId` is never re-used to mean a different TVC.
+
+The life cycle of each `GlobalId` follows several steps:
+
+0. The `GlobalId` is initially not yet bound.
+
+    In this state, the `GlobalId` cannot be used, as we do not yet know what its initial `since` will be.
+
+1. The `GlobalId` is bound to some definition, with an initially equal `since` and `upper`.
```

Comment:
Definition of what? A `GlobalId` can represent anything. Do we restrict ourselves to certain objects here (materialized views, source and sink definitions), or are we also including `GlobalId` values that represent users, roles, clusters, UDFs, and others.